api/v1alpha5/conversion.go

```diff
@@ -187,8 +188,22 @@ func (r *OpenStackMachineTemplateList) ConvertFrom(srcRaw ctrlconversion.Hub) er
 }
 
 func Convert_v1alpha8_OpenStackClusterSpec_To_v1alpha5_OpenStackClusterSpec(in *infrav1.OpenStackClusterSpec, out *OpenStackClusterSpec, s conversion.Scope) error {
-	// Our new flag has no equivalent in v1alpha5
-	return autoConvert_v1alpha8_OpenStackClusterSpec_To_v1alpha5_OpenStackClusterSpec(in, out, s)
+	err := autoConvert_v1alpha8_OpenStackClusterSpec_To_v1alpha5_OpenStackClusterSpec(in, out, s)
+	if err != nil {
+		return err
+	}
+
+	if in.Subnets != nil {
+		if len(in.Subnets) == 1 {
+			if err := Convert_v1alpha8_SubnetFilter_To_v1alpha5_SubnetFilter(&in.Subnets[0], &out.Subnet, s); err != nil {
+				return err
+			}
+		} else if len(in.Subnets) > 1 {
+			return fmt.Errorf("multiple Subnets not supported on v1alpha5 OpenStackClusterSpec")
+		}
```

Comment:
I'm _not sure_ we can fail conversions like this. Here I saw a lossy conversion in a similar case: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/16e467fb0aaa8a4fd15081481b9f39b476167215/api/v1alpha6/conversion.go#L526-L529

Same goes for the rest of conversions. Would love if @mdbooth or @EmilienM would chime in here.

Reply:
Nice catch. I'm incline to suggest to follow what has been done already and make it lossy (and document it)...

Reply:
In the case you linked, the `subnet` was transformed to `subnets` in the status as an anticipation of the dual-stack work, consequently there wouldn't actually be `subnets` with more than one subnet as we didn't accept that in the spec as well. I believe that is the reason @mdbooth  decided to only down convert one unique subnet. Let's wait for his input on this one.

With the current addition to the spec we might have a dual-stack cluster and if we try to convert it to older releases it would become single-stack. I believe this can cause inconsistencies across the cluster, for example some machines being dual-stack and others single-stack.

Reply:
I'm leaning towards the lie, e.g. return the first subnet in the list. My reasoning is this:
If we add a new field, the the conversion would be to just drop it, not error out.
Similarly, when we extend existing functionality, like with the `serverGroupID -> serverGroup`, we convert what we can but drop the rest. For example if a `serverGroup.name` is specified, we just drop this, but if we have `serverGroup.id` then we convert it.

For this reason I think it is fine to also drop the "additional" subnets when converting v1alpha8 -> v1alpha7.

Reply:
I would recommend reviewing [the guidance on converting fields to plural](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#making-a-singular-field-plural) in the API conventions.

My advice here, and the important things I think to consider are:
* APIs must continue to behave in the same way, for old versions
* Conversions should not return errors
* While the storage version is still alpha7
  * Writes to alpha8 must store their data in an annotation on the CRD on down convert, copy the value from index 0 to the old field
  * Reads from alpha8 must restore their data from the annotation, unless the field has since been changed by a write to alpha7
  * Writes to alpha7 will write without going through conversion webhook, but, mean that the user intended a singular value, so it's important once written to in alpha7, if I read alpha8, I get the singular value
  * Read from alpha7 will happen as it has before